internal/abi/prim/prim.go

```diff
@@ -25,10 +25,18 @@ type U32 uint32
 // U64 is an unsigned 64 bit integer.
 type U64 uint64
 
+// Pointer is the type of a pointer
+type Pointer[_ any] uint32
+
+// ToPointer turns an arbitrary pointer into a Pointer
+func ToPointer[T any](ptr *T) Pointer[T] {
+	return Pointer[T](uintptr(unsafe.Pointer(ptr)))
+}
+
 // Wstring is a header for a string.
 type Wstring struct {
-	Data uintptr
-	Len  uint
+	Data Pointer[U8]
```

Comment:
This is the only one I saw where the new type was different from the old one. 

Just working through this to make sure I understand: this could have been a `*prim.U8` before (a 32-bit value because it's a pointer), but was for whatever reason a `uintptr`, which is a 32-bit value in WASI because it's a 32-bit architecture?

Reply:
Yes.  `uintptr` is the integer version of a pointer.  TinyGo has a 32-bit wasm port, so `uintptr` was effectively a `uint32`, but Go is currently a weird 32/64-bit hybrid, so a `uintptr` is actually a `uint64`.  However, since `wasmtime` is a 32-bit wasm environment, we need to pass all the pointers as 32-bit things instead of 64-bit things.  Our `prim.Pointer` is effectively a `uint32`, so I decided to turn the `uintptr` into a `prim.Pointer[U8]`.

Reply:
This was helpful, thank you!